pypit/core/arextract.py

```diff
@@ -30,6 +30,934 @@
+        # this to work
+        temp = np.array([image[x1, y1] for (x1, y1) in zip(bigy.flatten(), fullspot.flatten())])
+        temp2 = np.reshape(weight.flatten() * temp, (nTrace, npix, tempx))
+        fextract = np.sum(temp2, axis=2)
+
+    # IDL version model functionality not implemented yet
+    # At the moment I'm not reutnring the f_ivar for the weight_image mode. I'm not sure that this functionality is even
+    # ever used
+
+    if(nTrace ==1):
+        fextract = fextract.reshape(npix)
+    return fextract.T
+
+
+def extract_boxcar(image,trace_in, radius, ycen = None):
+    """ Extract the total flux within a boxcar window at many positions. Based on idlspec2d/spec2d/extract_boxcar.pro
+
+    Parameters
+    ----------
+    image :   numpy float 2-d array [nspec, nspat]
+    trace :   Lower boundary of boxcar window (given as floating pt pixels) [nspec,nTrace]
+    radius :  boxcar radius (given as floating pt pixels)
+
+    Optional Parameters
+    -------------------
+    ycen :    Y positions corresponding to "trace" (expected as integers) [nspec,nTrace]
+
+    Returns
+    -------
+    fextract:   Extracted flux at positions within (trace +- radius, ycen) [nspec,nTrace]
+
+    Revision History
+    ----------------
+    24-Mar-1999  Written by David Schlegel, Princeton.
+    22-Apr-2018  Ported to python by Joe Hennawi, UCSB
+    """
+
+    trace = trace_in.T
+
+    if not (isinstance(radius,int) or isinstance(radius,float)):
+        raise ValueError('Boxcar radius must a be a floating point number')
+
+    dim = trace.shape
+    ndim = len(dim)
+    if (ndim == 1):
+        nTrace = 1
+        npix = dim[0]
+    else:
+        nTrace = dim[0]
+        npix = dim[1]
+
+    if ycen == None:
+        if ndim == 1:
+            ycen = np.arange(npix, dtype='int')
+        elif ndim == 2:
+            ycen = np.outer(np.ones(nTrace, dtype='int'), np.arange(npix, dtype='int'), )
+        else:
+            raise ValueError('trace is not 1 or 2 dimensional')
+
+    ycen_out = ycen.astype(int)
+    if np.size(trace) != np.size(ycen_out):
+        raise ValueError('Number of elements in trace and ycen must be equal')
+
+    left = trace - radius
+    right = trace + radius
+    fextract = extract_asymbox2(image, left, right, ycen_out)
+
+    return fextract
+
+
+def extract_optimal(waveimg, imgminsky, ivar, mask, oprof, skyimg, rn_img, box_radius, specobj):
+
+    nspat = imgminsky.shape[1]
+    nspec = imgminsky.shape[0]
+    nobj = len(specobjs)
+
+    spec_vec = np.arange(nspec)
+    spat_vec = np.arange(nspat)
+
+    var_no = np.abs(skyimg - np.sqrt(2.0) * rn_img) +rn_img**2
+
+    ispec, ispat = np.where(oprof > 0.0)
+    mincol = np.min(ispat)
+    maxcol = np.max(ispat) + 1
+    nsub = maxcol - mincol
+
+    mask_sub = mask[:,mincol:maxcol]
+    wave_sub = waveimg[:,mincol:maxcol]
+    ivar_sub = np.fmax(ivar[:,mincol:maxcol],0.0) # enforce positivity since these are used as weights
```

Comment:
According to the ```numpy``` docs, the inputs must have the
to have the same shape.  But I presume this is working ok.
I use ```np.maximum``` which is nearly the same but propogates
NaNs.  I guess you don't want that here.

Reply:
Or maybe you do given that you do a nansum below..